MAINT trigger lock file update in PR with a comment

Charlie-XIAO opened the pull request:
Towards #28274.

This PR adds a workflow that can be triggered by specific PR comments to update environments and lock files in that PR. The comment should start with `please update lock-files`, and then followed by optional square brackets where `tag` corresponds to `--select-tag`, `build` corresponds to `--select-build`, and `-build` corresponds to `--skip-build` in `build_tools/update_environments_and_lock_files.py`. A special case is `please update lock-files all` which will run the script without any arguments. Examples are:

```
please update lock-files all
please update lock-files [tag main-ci]
please update lock-files [build doc] [-build doc_dependencies]
```

Invalid examples that will not trigger an update include:

```
please update lock-files                   # Not specifying any option, and not explicitly specifying all
please update lock-files [ttag main-ci]    # Typo in options
```

You may also check https://github.com/Charlie-XIAO/github-actions-test/pull/5 for a brief demo, since I think this PR cannot be tested unless merged into `main`.

Charlie-XIAO committed:
MAINT trigger lock file update in PR with a comment

Charlie-XIAO committed:
CodeQL: potential code injection

Charlie-XIAO commented on .github/workflows/update-lock-files-pr.yml:
```diff
@@ -0,0 +1,110 @@
+            echo "select_tag=$select_tag" >> $GITHUB_OUTPUT
+            echo "select_build=$select_build" >> $GITHUB_OUTPUT
+            echo "skip_build=$skip_build" >> $GITHUB_OUTPUT
+          fi
```
https://github.com/scikit-learn/scikit-learn/pull/29505/commits/ba2eb2692adc55b0025ba995a7b74ed442671e48

lesteve commented:
Thanks for initiating this! Off the top of my head the thing that need to be checked on your toy repo:
- we want a bot commit to be added into the contributor branch so I guess care needs to be taken about permission and/or token. Looking at how conda-forge please rerender may be useful to see how they do it.
- if you test on your fork you should check on a PR from an external contributor, I can do a test
- trigger: I guess anyone that comments should be able to trigger lock file update, i.e. we don't want to restrict this special power only to maintainers.
- not really crucial right now, but I would not add any special parsing logic and pass directly the args to the command something like `@scikit-learn-bot update lock-file --select-build `doc-min-dependencies`

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

Charlie-XIAO commented:
Looks much more complicated than I imagined 🤔 need to take some time to investigate

lesteve commented:
> Looks much more complicated than I imagined 

Maybe I should not have done a brain dump of all the things that I had in mind :wink:

Charlie-XIAO commented:
No that's really helpful, otherwise I could have gone in the wrong direction 😁

Charlie-XIAO committed:
simplify parsing; tackle PRs made from external forks

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

Charlie-XIAO commented:
> we want a bot commit to be added into the contributor branch so I guess care needs to be taken about permission and/or token. Looking at how conda-forge please rerender may be useful to see how they do it. Actually looks like this is more sophisticated than what I thought they have a webapp according to https://conda-forge.org/docs/maintainer/infrastructure/#admin-web-services. We should probably aim for something lighter.

True I don't think we can do something like that. I looked at their infrastructure and there's the docker holding the service and several dispatches. As for the permission, I think being able to write to a PR made from a fork that does not belong to a maintainer is enough? I think this can be illustrated in the next two points.

> If you test on your fork you should check on a PR from an external contributor, I can do a test with my account or you can create a toy account. For example personally I have github.com/test-lesteve for this kind of use cases

I created @Charlie-XIAO-qq :) Opened Charlie-XIAO/github-actions-test#8 as a demo.

> Trigger: I guess anyone that comments should be able to trigger lock file update, i.e. we don't want to restrict this special power only to maintainers.

Again illustrated in Charlie-XIAO/github-actions-test#8.

> Not really crucial right now, but I would not add any special parsing logic and pass directly the args to the command something like `@scikit-learn-bot update lock-file --select-build doc-min-dependencies` (hmmm I guess you may need to find a way to specify the marker as well? like `[doc build]` or `[scipy-dev]` so maybe you do need some king of simple args parsing) ...

My original considerations are that if there are clear typos then we can skip all those environment setup steps, but if this is not needed actually the parsing logic becomes much simpler. Re-implemented in https://github.com/scikit-learn/scikit-learn/pull/29505/commits/f921b90ab61b7f0031005fcc3f394b35b46195c2.

lesteve commented:
Nice, I thought some special mechanism would be needed so that `scikit-learn-bot` can push into PR author's branch, but actually it seems like this is not needed.

Do you know why this is not needed? Maybe I guess the token has by default the necessary write permissions to push into the contributor branch?

Just out of curiosity, I tried with the "Allow edit by maintainers" box unticked when creating the PR and the workflow fails but that's completely acceptable: https://github.com/Charlie-XIAO/github-actions-test/pull/9.

Maybe not in this PR, but it would be very nice if there was some kind of feed-back to indicate that something did actually happen after the comment is made, ideally with a CI status or alternatively as an edit to the comment with a link to the GHA log ? My thinking is that generating the lock-files can take a few minutes and without feed-back it's quite easy to think that nothing has happened ...

Charlie-XIAO commented:
> Nice, I thought some special mechanism would be needed so that `scikit-learn-bot` can push into PR author's branch, but actually it seems like this is not needed.
>
> Do you know why this is not needed? Maybe I guess the token has by default the necessary write permissions to push into the contributor branch?

I'm also pretty surprised that it works directly. Since `scikit-learn-bot` is clearly not a maintainer of my [Charlie-XIAO/github-actions-test](https://github.com/Charlie-XIAO/github-actions-test) repo I believe its not related to the configuration of the bot. I think it's in the `actions/checkout` step, where the default token is `github.token` (equivalent to `GITHUB_TOKEN`) which have read/write permissions to pull requests ([GitHub docs](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token)).

> Maybe not in this PR, but it would be very nice if there was some kind of feed-back to indicate that something did actually happen after the comment is made, ideally with a CI status or alternatively as an edit to the comment with a link to the GHA log ? My thinking is that generating the lock-files can take a few minutes and without feed-back it's quite easy to think that nothing has happened ...

Totally agreed. I see the [peter-evans/create-or-update-comment](https://github.com/peter-evans/create-or-update-comment) action that can be used to create/update comments and reactions, and I believe we also have something that create/update comments in the lint workflow. Which one do you think I should use?

lesteve commented:
A few comments:
- can you update the permissions so that it only has the necessary permissions? This workflow can push into a contributor's branch based on anyone's comment which seems powerful and the fact that we were surprised it was possible makes me a bit worried about other potential suprises
- I think the bash argument logic is a bit fragile, maybe it could be a Python script even if it is 10 lines that mostly call `build_tools/update_environment_and_lock_files.py` and do a commit

To be honest maybe we should be tweaking the default permissions to be read by default, although this may break some existing workflows not sure, [settings page](https://github.com/scikit-learn/scikit-learn/settings/actions)
![image](https://github.com/user-attachments/assets/c65a27dd-4f72-47d4-a5e6-4073fb9578b2)

lesteve commented on .github/workflows/update-lock-files-pr.yml:
```diff
@@ -0,0 +1,88 @@
+          esac
+
+          git config --global user.name "scikit-learn-bot"
+          git config --global user.email "tjpfdev@gmail.com"
```
this is actually one of Thomas' email address who created the Github scikit-learn-bot user. Probably you can use noreply@github.com even if on github it will look like a commit from https://github.com/web-flow e.g. as in this commit https://github.com/scikit-learn/scikit-learn/commit/4a62414629ad364a986b1688ee6f7bc54958e149

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

Charlie-XIAO committed:
resolve conversations

Charlie-XIAO commented:
See https://github.com/Charlie-XIAO/github-actions-test/pull/11 for the new demo where I addressed most of the suggestions in https://github.com/scikit-learn/scikit-learn/pull/29505#issuecomment-2248245812.

Regarding the permissions, I've not changed the scikit-learn main repo but I believe the additional `contents: write` is enough according to experiments in my demo repository.

As for the Python script, it works fine https://github.com/scikit-learn/scikit-learn/pull/29505/commits/bdc234d3f3509e1e363e56509c357524ad452fc3 but let me know if this is not what you meant.

lesteve commented on build_tools/update_environments_and_lock_files_pr.py:
```diff
@@ -0,0 +1,47 @@
+    elif tag == "arm":
+        marker = "[cirrus arm] "
+    else:
+        marker = ""
```
I think by default this updates all the lock-files so all the markers should be present `[doc build] [scipy-dev] [cirrus arm]`. Not sure this will happen that often in practice though that you want to update all the lock-files at once ...

lesteve commented on build_tools/update_environments_and_lock_files_pr.py:
```diff
@@ -0,0 +1,47 @@
+
+    # Extract the tag from the args (if any)
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--select-tag", default=None)
```
I think it would be great to allow to pass specific marker, e.g. via a `--commit-marker` option?

I can imagine that one of the use case is when you want to update only the `doc` and `doc-min-dependencies` lock-files in which case you would use something like this:
```
@scikit-learn-bot please update lock-file --select-build doc --commit-marker '[doc build]` 
```

lesteve reviewed (commented):
This is progressing smoothly thanks!

Still a few comments and then afterwards we can try to find a second reviewer (maybe Adrin since he is the one who created the original issue).

lesteve commented on build_tools/update_environments_and_lock_files_pr.py:
I forgot to say not sure about the file name here, it seems to close to the `build_tools/update_environments_and_lock_files.py`, I don't have a great suggestion though ... maybe put the variation at the beginning like `build_tools/on_pr_comment_update_environment_and_lock_files.py` (a bit long but oh well it's only for the CI so maybe good enough)

Alternatively we could use a single file `build_tools/update_environment_and_lock_files.py` and add the git functionality behind an option `--bot-commit` or something, I am not sure. This could be handy to actually figure out the commit marker from the selected builds, but probably we can leave this for later.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

Charlie-XIAO committed:
resolve conversations

Charlie-XIAO committed:
rename file

lesteve commented on .github/workflows/update-lock-files-pr.yml:
```diff
@@ -0,0 +1,48 @@
+          conda config --set solver libmamba
+          conda install -c conda-forge "$(get_dep conda-lock min)" -y
+
+          python build_tools/update_environments_and_lock_files_pr.py
```
You need this, don't you?
```suggestion
          python build_tools/on_pr_comment_update_environments_and_lock_files.py
```

Charlie-XIAO replied:
Yes thanks for the reminder. I'm still playing with my demo and there seems to be something else going wrong as well.

lesteve commented:
ping @adrinjalali since you created the original issue #28274 you may have some interest in moving forward your "lock-file update as a PR comment" idea.

It would be certainly good to get a second pair of eyes on this.

The main improvement in a separate PR would be to add a CI status to indicate something is happening personally I prefer a CI status for this which I find less noisy than a comment even if the same comment is being edited like in https://github.com/peter-evans/create-or-update-comment.

I guess it's probably doable to use the [github REST API](https://docs.github.com/en/rest/commits/statuses?apiVersion=2022-11-28#create-a-commit-status) for this, maybe gh CLI (but I haven't found it if it's possible in 5 minutes of googling) or at last resort an external action like https://github.com/myrotvorets/set-commit-status-action. I am a bit wary of using external action if we can avoid it, since we got the security report a month or so ago and I got more aware of a few ways Github actions can be used for abuse. It feels like here the worst that could happen is that the external action use the token with write rights to push into the PR contributor branch, which seems like a bit bad.

Charlie-XIAO committed:
fix dealing with extra commit marker

Charlie-XIAO commented:
Here is the latest demo: https://github.com/Charlie-XIAO/github-actions-test/pull/14

To deal with `--commit-marker` the simplest way is to add an unused option in `build_tools/update_environments_and_lock_files.py` but not so elegant, so I still handled it in the PR comment script with some additional complexity in reassembling the command line.

lesteve reviewed (approved):
This feels good enough for me to merge, try to use it and improve it where needed. This would be already super useful, thanks a lot for your work @Charlie-XIAO!

A few ideas for improvements for further PRs:
- add a commit status to indicate that something is happening and lock-files are being regenerated
- maybe a simpler way for the comment syntax? It feels a bit weird to write pieces of CLI commands inside a PR comment, so could it be simpler? This also means having some more complicated custom logic in `build_tools/on_pr_comment_update_environments_and_lock_files.py` and this is maybe not worth it.

lesteve commented:
@Charlie-XIAO I forgot to say, this would be nice to know if you have tested this on your scikit-learn fork rather than only in your demo repo. This is so easy to get some details wrong with this kind of CI stuff (definitely speaking from experience :sweat_smile:)  ...

adrinjalali commented on .github/workflows/update-lock-files-pr.yml:
```diff
@@ -0,0 +1,48 @@
+  issue_comment:
+    types: [created]
+
+permissions:
+  contents: write
```
this means this action has write permissions on the repo, as a maintainer, which is pretty wild, so we should make sure we have the right safety mechanisms here. So the first thing to check is, if a user w/o write permissions submits a PR which modifies this file, or the corresponding python file doing the changes, is this action then using the PR's version or the version from `main`?

Charlie-XIAO replied:
I believe changes are always pushed to the PR branch instead of `main`, because we are getting the repository and ref to checkout using the following mechanism:

```bash
pr_info=$(gh pr view ${{ github.event.issue.number }} --repo ${{ github.repository }} --json headRefName,headRepository,headRepositoryOwner)
pr_head_ref=$(echo "$pr_info" | jq -r '.headRefName')
pr_head_repository=$(echo "$pr_info" | jq -r '.headRepositoryOwner.login + "/" + .headRepository.name')
```

And BTW actually all PRs have read and write permissions to this repo currently ([ref](https://github.com/scikit-learn/scikit-learn/settings/actions)); these lines are added because we may want to disable the global write permissions some day https://github.com/scikit-learn/scikit-learn/pull/29505#issuecomment-2248245812.

adrinjalali replied:
What I mean is, if the user changes the code we have here, they can change it to targeting `main` and write to `main` or one of our branches on the main repo.

adrinjalali commented on .github/workflows/update-lock-files-pr.yml:
```diff
@@ -0,0 +1,48 @@
+        run: |
+          source build_tools/shared.sh
+          source $CONDA/bin/activate
+          conda install -n base conda conda-libmamba-solver -y
+          conda config --set solver libmamba
+          conda install -c conda-forge "$(get_dep conda-lock min)" -y
```
micromamba is gonna be a lot faster here. But well...

Charlie-XIAO replied:
I'm copying these from `update-lock-files.yml`; if there's something faster I guess we can update both in a follow-up PR?

adrinjalali commented on build_tools/on_pr_comment_update_environments_and_lock_files.py:
```diff
@@ -0,0 +1,67 @@
+    )
+    execute_command('git config --global user.name "scikit-learn-bot"')
+    execute_command('git config --global user.email "noreply@github.com"')
+    execute_command("git add -A")
+    execute_command(f'git commit -m "{marker}Update lock files"')
+    execute_command("git push")
```
maybe here we should make sure we're not pushing to the main repo? As in, check `remote` and make sure it's not `scikit-learn` and it's the same as the users?

Charlie-XIAO replied:
IMHO it's not necessary because I think we can trust the official `actions/checkout` that when pushing it would always respect the repository and ref we checked out, which as mentioned in https://github.com/scikit-learn/scikit-learn/pull/29505#discussion_r1713170424 will (theoretically) be the PR branch in the contributor's fork. Perhaps we can instead check that `pr_head_repository` and `pr_head_ref` are not `scikit-learn/scikit-learn` and `main` in an earlier step?

adrinjalali reviewed (commented):
I would have liked a comment from the bot once this is done, but it seems @lesteve doesn't like comments.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

Charlie-XIAO committed:
[DANGER] temporarily remove repo check for testing on my fork

Charlie-XIAO commented:
Sorry for the long delay.

@adrinjalali Thanks for your review and I've made responses to your review comments. In general I agree that we should be extra careful about workflows that have write permissions (especially to contents), but IMHO it's enough to make sure we checked out the correct branch with `@actions/checkout`.

> I forgot to say, this would be nice to know if you have tested this on your scikit-learn fork rather than only in your demo repo. This is so easy to get some details wrong with this kind of CI stuff (definitely speaking from experience 😅) ...

@lesteve I've changed the default branch to this PR branch in my fork (so that the workflow can be triggered). See https://github.com/Charlie-XIAO/scikit-learn/pull/1 (in particular, [this workflow run](https://github.com/Charlie-XIAO/scikit-learn/actions/runs/10347651692) and [this commit](https://github.com/Charlie-XIAO/scikit-learn/pull/1/commits/8bff0f9bbdc535f70b620e38c54647ffdcf330d4)).

adrinjalali commented:
One thing which is still not clear to me, is when somebody opens a PR which changes these files, then we will be vulnerable to attacks. That's why here (https://github.com/scikit-learn/scikit-learn/blob/main/.github/workflows/lint.yml) I download from the main repo, and it's a `pull_request_target`. I'm not saying here we're vulnerable, what I'm saying is that I'd like us to check we're not.

Charlie-XIAO commented:
According to the second note in this [GitHub docs page](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#issue_comment), I think the workflow only uses the version on `main` (so that changes in a PR won't have an effect unless merged into `main`)? For instance in this PR if we do a `@scikit-learn-bot update lock-files` comment nothing would happen because the workflow file does not exist on `main` yet.

adrinjalali commented:
The documentation says:

> Note: This event will only trigger a workflow run if the workflow file is on the default branch.

But it doesn't say it always gets the version from the default branch. In many cases having the file in the default branch "creates" the workflow, but the code of the workflow is fetched from the current PR. Could you double check this?

Charlie-XIAO commented:
Okay lemme check in my test repo.

Charlie-XIAO commented:
@adrinjalali I'm not 100% sure by it seems that changes in the workflow file won't be applied, but changes in the Python script that it calls will; see https://github.com/Charlie-XIAO/github-actions-test/pull/14#issuecomment-2284002267 (and the commit above and the workflow it references). Would it mean that we should avoid using a Python script here?

Also I think we have branch protection rules for `main`, so would that suffice (I mean, it will deny a direct push like in the workflow)?

adrinjalali commented:
> Would it mean that we should avoid using a Python script here?

That means we should manually download / use the python script from the main branch the same way done in the linter bot commenter file.

Charlie-XIAO committed:
download scripts from main for safety

Charlie-XIAO committed:
[DANGER] testing purposes only

Charlie-XIAO commented:
Ah okay I finally got the point 🤣 forgive me for being dumb.

Now I updated and tested in https://github.com/Charlie-XIAO/scikit-learn/pull/1 it works fine. Note that there are a few modifications for testing purposes that need to be reverted.

adrinjalali reviewed (commented):
Nice, now this feels better. Could you please also add something in our docs (beside where we have all the especial git commit messages I guess)? Other than the test purpose changes, LGTM.

Charlie-XIAO committed:
documentation for the workflow

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into comment-update-lock

adrinjalali commented:
Let's see how it works then

lesteve commented:
Nice to see this has been merged! Just curious, has this already been tested in a scikit-learn PR?

Charlie-XIAO commented:
Nope but I think I'm going to make a PR updating to sphinx-gallery 0.17.1 within a few days where we can try that workflow :)

Charlie-XIAO commented:
Oops I forgot to remove the test-only changes before merging so the workflow is failing: https://github.com/scikit-learn/scikit-learn/actions/runs/10499866845/job/29087364777

By the way https://github.com/scikit-learn/scikit-learn/pull/29700 is the PR I'm going to test the workflow.
